File: benchmarks/mason/GameOfLife.java
import sim.engine.SimState;
import sim.engine.Steppable;
import sim.field.grid.IntGrid2D;

/**
 * Conway's Game of Life, B3/S23 on a Moore torus, updated synchronously.
 *
 * Follows the shape of MASON's own cellular automaton tutorial. One Steppable owns the whole
 * `IntGrid2D` and writes into a second grid copied back at the end of the sweep. MASON shuffles
 * agents that share a time and an ordering, and a Steppable per cell would be asynchronous.
 */
public class GameOfLife extends HenadModel {
    static final int DEAD = 0;
    static final int ALIVE = 1;

    /** Reads `state` rather than closing over a model, so one instance serves every run. */
    private static final Steppable SWEEP = new Steppable() {
        public void step(SimState state) {
            ((GameOfLife) state).sweep();
        }
    };

    public final IntGrid2D grid;
    private final IntGrid2D next;
    private final double density;
    private final int[][] live;

    /** `live` places an exact set of `(x, y)` cells for a gate scenario, otherwise `density` fills at random. */
    public GameOfLife(long seed, int width, int height, double density, int[][] live) {
        super(seed);
        this.grid = new IntGrid2D(width, height, DEAD);
        this.next = new IntGrid2D(width, height, DEAD);
        this.density = density;
        this.live = live;
    }

    public void start() {
        super.start();
        populate();
        schedule.scheduleRepeating(SWEEP);
    }

    private void populate() {
        final int w = grid.getWidth();
        final int h = grid.getHeight();
        grid.setTo(DEAD);
        if (live != null) {
            for (int[] cell : live) {
                grid.field[cell[0]][cell[1]] = ALIVE;
            }
            return;
        }
        for (int x = 0; x < w; x++) {
            for (int y = 0; y < h; y++) {
                grid.field[x][y] = random.nextDouble() < density ? ALIVE : DEAD;
            }
        }
    }

    private void sweep() {
        final int w = grid.getWidth();
        final int h = grid.getHeight();
        final int[][] cur = grid.field;
        final int[][] out = next.field;
        for (int x = 0; x < w; x++) {
            // stx wraps by one width. A Moore radius of 1 never asks for more.
            final int[] left = cur[grid.stx(x - 1)];
            final int[] mid = cur[x];
            final int[] right = cur[grid.stx(x + 1)];
            final int[] col = out[x];
            for (int y = 0; y < h; y++) {
                final int up = grid.sty(y - 1);
                final int down = grid.sty(y + 1);
                final int alive = left[up] + left[y] + left[down]
                        + mid[up] + mid[down]
                        + right[up] + right[y] + right[down];
                col[y] = mid[y] == ALIVE
                        ? (alive == 2 || alive == 3 ? ALIVE : DEAD)
                        : (alive == 3 ? ALIVE : DEAD);
            }
        }
        grid.setTo(next);
    }

    public int population() {
        return grid.getWidth() * grid.getHeight();
    }

    /** Rows ascending in y, each row ascending in x, as the fixture format wants. */
    public String[] bitmap() {
        final int w = grid.getWidth();
        final int h = grid.getHeight();
        final String[] rows = new String[h];
        final StringBuilder row = new StringBuilder(w);
        for (int y = 0; y < h; y++) {
            row.setLength(0);
            for (int x = 0; x < w; x++) {
                row.append((char) ('0' + grid.field[x][y]));
            }
            rows[y] = row.toString();
        }
        return rows;
    }
}

File: benchmarks/krabmaga/src/models/game_of_life.rs
//! Conway's Game of Life, B3/S23 on a Moore torus, updated synchronously.
//!
//! Follows the shape of krABMaga's own `forestfire` example. One agent sweeps the whole grid,
//! reading the field's read buffer and writing its write buffer, and `Schedule::step` swaps the
//! two once every cell has been written. No cell sees a neighbour that has already moved on.

use std::any::Any;

use krabmaga::engine::agent::Agent;
use krabmaga::engine::fields::dense_number_grid_2d::DenseNumberGrid2D;
use krabmaga::engine::fields::field::Field;
use krabmaga::engine::location::Int2D;
use krabmaga::engine::schedule::Schedule;
use krabmaga::engine::state::State;
use krabmaga::rand::{Rng, SeedableRng};
use krabmaga::rand_pcg::Pcg64;

use crate::models::wrap;

const DEAD: u8 = 0;
const ALIVE: u8 = 1;

#[derive(Clone)]
struct Sweeper;

impl Agent for Sweeper {
    fn step(&mut self, state: &mut dyn State) {
        let model = state.as_any().downcast_ref::<GameOfLife>().expect("Game of Life state");
        let (width, height) = (model.width, model.height);

        for x in 0..width {
            for y in 0..height {
                let mut alive = 0u8;
                for dx in -1..=1 {
                    for dy in -1..=1 {
                        if dx == 0 && dy == 0 {
                            continue;
                        }
                        let loc = Int2D {
                            x: wrap(x + dx, width),
                            y: wrap(y + dy, height),
                        };
                        alive += model.cells.get_value(&loc).unwrap_or(DEAD);
                    }
                }
                let here = Int2D { x, y };
                let cell = model.cells.get_value(&here).unwrap_or(DEAD);
                let next = match (cell, alive) {
                    (ALIVE, 2..=3) | (DEAD, 3) => ALIVE,
                    _ => DEAD,
                };
                model.cells.set_value_location(next, &here);
            }
        }
    }
}

pub struct GameOfLife {
    pub width: i32,
    pub height: i32,
    cells: DenseNumberGrid2D<u8>,
    density: f64,
    /// An exact set of live cells for a gate scenario, in place of the random fill.
    live: Option<Vec<(i32, i32)>>,
    seed: u64,
}

impl GameOfLife {
    pub fn new(width: i32, height: i32, density: f64, seed: u64) -> Self {
        GameOfLife {
            width,
            height,
            cells: DenseNumberGrid2D::new(width, height),
            density,
            live: None,
            seed,
        }
    }

    pub fn with_live(width: i32, height: i32, live: &[(i32, i32)]) -> Self {
        let mut model = GameOfLife::new(width, height, 0.0, 0);
        model.live = Some(live.to_vec());
        model
    }

    pub fn population(&self) -> u64 {
        self.width as u64 * self.height as u64
    }

    /// Before the first step the grid still sits in the write buffer, since `Schedule::step` is
    /// what publishes it.
    fn cell(&self, x: i32, y: i32) -> u8 {
        let loc = Int2D { x, y };
        self.cells
            .get_value(&loc)
            .or_else(|| self.cells.get_value_unbuffered(&loc))
            .unwrap_or(DEAD)
    }

    /// Rows ascending in y, each row ascending in x, as the fixture format wants.
    pub fn bitmap(&self) -> Vec<String> {
        (0..self.height)
            .map(|y| (0..self.width).map(|x| char::from(b'0' + self.cell(x, y))).collect())
            .collect()
    }
}

impl State for GameOfLife {
    fn init(&mut self, schedule: &mut Schedule) {
        let mut rng = Pcg64::seed_from_u64(self.seed);
        for x in 0..self.width {
            for y in 0..self.height {
                let state = match &self.live {
                    Some(live) => {
                        if live.contains(&(x, y)) {
                            ALIVE
                        } else {
                            DEAD
                        }
                    }
                    None => {
                        if rng.random::<f64>() < self.density {
                            ALIVE
                        } else {
                            DEAD
                        }
                    }
                };
                self.cells.set_value_location(state, &Int2D { x, y });
            }
        }
        schedule.schedule_repeating(Box::new(Sweeper), 0.0, 0);
    }

    fn update(&mut self, _step: u64) {
        self.cells.lazy_update();
    }

    fn reset(&mut self) {
        self.cells = DenseNumberGrid2D::new(self.width, self.height);
    }

    fn as_any(&self) -> &dyn Any {
        self
    }

    fn as_any_mut(&mut self) -> &mut dyn Any {
        self
    }

    fn as_state(&self) -> &dyn State {
        self
    }

    fn as_state_mut(&mut self) -> &mut dyn State {
        self
    }
}
